jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/NodeRecordTask.java

```diff
@@ -0,0 +1,166 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.github.jbellis.jvector.graph.disk;
+
+import io.github.jbellis.jvector.disk.ByteBufferIndexWriter;
+import io.github.jbellis.jvector.graph.ImmutableGraphIndex;
+import io.github.jbellis.jvector.graph.disk.feature.Feature;
+import io.github.jbellis.jvector.graph.disk.feature.FeatureId;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.function.IntFunction;
+
+/**
+ * A task that builds a single node's L0 record in memory.
+ * <p>
+ * This task is designed to be executed in a thread pool, with each worker thread
+ * owning its own ImmutableGraphIndex.View for thread-safe neighbor iteration.
+ */
+class NodeRecordTask implements Callable<NodeRecordTask.Result> {
+    private final int newOrdinal;
+    private final OrdinalMapper ordinalMapper;
+    private final ImmutableGraphIndex graph;
+    private final ImmutableGraphIndex.View view;
+    private final List<Feature> inlineFeatures;
+    private final Map<FeatureId, IntFunction<Feature.State>> featureStateSuppliers;
+    private final int recordSize;
+    private final long fileOffset;
+    private final ByteBuffer buffer;
+
+    /**
+     * Result of building a node record.
+     */
+    static class Result {
+        final int newOrdinal;
+        final long fileOffset;
+        final ByteBuffer data;
+
+        Result(int newOrdinal, long fileOffset, ByteBuffer data) {
+            this.newOrdinal = newOrdinal;
+            this.fileOffset = fileOffset;
+            this.data = data;
+        }
+    }
+
+    NodeRecordTask(int newOrdinal,
+                   OrdinalMapper ordinalMapper,
+                   ImmutableGraphIndex graph,
+                   ImmutableGraphIndex.View view,
+                   List<Feature> inlineFeatures,
+                   Map<FeatureId, IntFunction<Feature.State>> featureStateSuppliers,
+                   int recordSize,
+                   long fileOffset,
+                   ByteBuffer buffer) {
+        this.newOrdinal = newOrdinal;
+        this.ordinalMapper = ordinalMapper;
+        this.graph = graph;
+        this.view = view;
+        this.inlineFeatures = inlineFeatures;
+        this.featureStateSuppliers = featureStateSuppliers;
+        this.recordSize = recordSize;
+        this.fileOffset = fileOffset;
+        this.buffer = buffer;
+    }
+
+    @Override
+    public Result call() throws Exception {
+        // Writer automatically clears buffer on construction
+        var writer = new ByteBufferIndexWriter(buffer);
```

Comment:
Why not make the `ByteBufferIndexWriter` thread local and avoid recreating it on every call?